### PR TITLE
Splitting per-workgroup params from per-dispatch params.

### PR DESCRIPTION
**Pull request by benvanik**
This allows for sharing the dispatch state across all workgroups.
The states are now packed such that they fit in single cache lines.

### CODE REVIEW

**Review by MaheshRavishankar — APPROVED** [review]

File: iree/hal/local/local_executable.c [hunk 1]
```diff
@@ -52,44 +52,52 @@ iree_hal_local_executable_t* iree_hal_local_executable_cast(
...
     const iree_hal_executable_dispatch_state_v0_t* dispatch_state,
-    iree_byte_span_t local_memory) {
+    uint32_t processor_id, iree_byte_span_t local_memory) {
   IREE_TRACE_ZONE_BEGIN(z0);
   // TODO(benvanik): annotate with executable name to calculate total time.
 
-  const iree_hal_vec3_t workgroup_count = dispatch_state->workgroup_count;
+  const uint32_t workgroup_count_x = dispatch_state->workgroup_count_x;
+  const uint32_t workgroup_count_y = dispatch_state->workgroup_count_y;
+  const uint32_t workgroup_count_z = dispatch_state->workgroup_count_z;
```
Comment (ScottTodd) [inline iree/hal/local/local_executable.c:73]: (see my other comment) This then pulls the `uint16_t` out as a `uint32_t`?

Comment (benvanik) [reply iree/hal/local/local_executable.c:73]: that's fine - the uint16_t is a compression for the in memory structure, not a requirement that it must be in a 16 bit register when used.

File: iree/hal/local/executable_library.h [hunk 2]
```diff
@@ -258,56 +258,91 @@ typedef struct iree_hal_executable_environment_v0_t {
...
-  // Total workgroup count for the dispatch. This is sourced from either the
-  // original dispatch call (for iree_hal_command_buffer_dispatch) or the
-  // indirection buffer (for iree_hal_command_buffer_dispatch_indirect).
-  iree_hal_vec3_t workgroup_count;
   // Workgroup size chosen for the dispatch. For compilation modes where the
   // workgroup size is constant this may be ignored.
-  iree_hal_vec3_t workgroup_size;
+  uint32_t workgroup_size_x;
+  uint32_t workgroup_size_y;
+  uint16_t workgroup_size_z;
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/hal/local/executable_library.h:270]: Good q: Z is used as the coarsest distribution dimension and there's rarely need to go larger than ~16 (divisions, not bits), and many cases it's just 1 - given that we should be targeting ~100's of workgroups putting all of the distribution on Z would be bad.

The thing that makes limiting here fine is that all practical grid dispatch systems also have this limitation - including the latest and greatest nvidia parts - so we need to be aware of the 16-bit Z limit on GPUs anyway: https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxComputeWorkGroupCount[2]&platform=all

Basically, if you need more than 65k workgroups along Z you're doing something wrong. Maybe one day if there's a UMA processor with >65k cores that we want to coarsely distribute across it'd be an issue but I think the speed of light prevents that from being the case :)

I'll add a note to the header.

[the remark this answers is not in the document]
Comment (ScottTodd) [reply iree/hal/local/executable_library.h:270]: 👍 sgtm, about what I figured. Thanks for the explanation!
